sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java

```diff
@@ -234,96 +263,178 @@ public String serializeList(List<?> list, CollectionFormat format) {
+            .visibility(PropertyAccessor.GETTER, JsonAutoDetect.Visibility.NONE)
+            .visibility(PropertyAccessor.IS_GETTER, JsonAutoDetect.Visibility.NONE);
+
         return mapper;
     }
 
     private JavaType createJavaType(Type type) {
-        JavaType result;
         if (type == null) {
-            result = null;
+            return null;
         } else if (type instanceof JavaType) {
-            result = (JavaType) type;
+            return (JavaType) type;
         } else if (type instanceof ParameterizedType) {
-            final ParameterizedType parameterizedType = (ParameterizedType) type;
-            final Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
-            JavaType[] javaTypeArguments = new JavaType[actualTypeArguments.length];
-            for (int i = 0; i != actualTypeArguments.length; i++) {
-                javaTypeArguments[i] = createJavaType(actualTypeArguments[i]);
-            }
-            result = mapper
-                .getTypeFactory().constructParametricType((Class<?>) parameterizedType.getRawType(), javaTypeArguments);
+            return typeToJavaTypeCache.computeIfAbsent(type, t -> {
+                final ParameterizedType parameterizedType = (ParameterizedType) type;
+                final Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
+                JavaType[] javaTypeArguments = new JavaType[actualTypeArguments.length];
+                for (int i = 0; i != actualTypeArguments.length; i++) {
+                    javaTypeArguments[i] = createJavaType(actualTypeArguments[i]);
+                }
+
+                return mapper.getTypeFactory().constructParametricType((Class<?>) parameterizedType.getRawType(),
+                    javaTypeArguments);
+            });
         } else {
-            result = mapper
-                .getTypeFactory().constructType(type);
+            return typeToJavaTypeCache.computeIfAbsent(type, t -> mapper.getTypeFactory().constructType(t));
         }
-        return result;
     }
 
+    /*
+     * Internal helper class that helps manage converting headers into their header collection.
+     */
+    private static final class HeaderCollectionHandler {
+        private final String prefix;
+        private final int prefixLength;
+        private final Map<String, String> values;
+        private final Field declaringField;
+
+        HeaderCollectionHandler(String prefix, Field declaringField) {
+            this.prefix = prefix;
+            this.prefixLength = prefix.length();
+            this.values = new HashMap<>();
+            this.declaringField = declaringField;
+        }
+
+        boolean headerStartsWithPrefix(String headerName) {
+            return headerName.startsWith(prefix);
+        }
+
+        void addHeader(String headerName, String headerValue) {
+            values.put(headerName.substring(prefixLength), headerValue);
```

Comment:
Should there be a check for `headerName.length() >= prefixLength`?

Reply:
This is already checked for this by the caller, validating that the header name `startsWith` the prefix.